src/header/internals/item.rs

```diff
@@ -82,6 +82,14 @@ impl Item {
         }
         self.typed.get_mut(tid).map(|typed| unsafe { typed.downcast_mut_unchecked() })
     }
+
+    pub fn into_typed<H: Header>(self) -> Option<H> {
```

Comment:
Adding this to the `impl Header + Send + Sync` should do it:
```rust
#[inline]
unsafe fn downcast_unchecked<T: 'static>(self: Box<Self>) -> T {
    *(mem::transmute::<*mut _, (*mut (), *mut ())>(Box::into_raw(self)).0 as *mut T)
}
```

Then you can call `val.downcast_unchecked()` after having gotten the type from the TypeId.